dspace/config/spring/api/scripts.xml

```diff
@@ -4,6 +4,10 @@
     xsi:schemaLocation="http://www.springframework.org/schema/beans
            http://www.springframework.org/schema/beans/spring-beans-2.5.xsd">
 
+    <bean id="dspaceRunnableThreadExecutor" class="org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor">
+        <property name="corePoolSize" value="5"/>
```

Comment:
**Could you add a comment to explain why this is set to 5?**

There is no particular reason that this is set to 5, it’s just a default to not allow too many threads to be working at the same time and potentially cause memory issues or other performance issues.  This was [previously also set to 5](https://github.com/DSpace/DSpace/blob/05231676253d397262569f1f5386b6b297547b22/dspace/config/spring/rest/scripts.xml#L6), so we kept this as is

**I’m also curious why the production version of `dspaceRunnableThreadExecutor` is of type `ThreadPoolTaskExecutor`, while all the tests use `SyncTaskExecutor`. Shouldn’t the tests use the same type of TaskExecutor as the main code? If not, I think we should add an inline comment to describe the reason for the difference (it’s not obvious to me).**

The DSpaceRunnableThreadExecutor is responsible for executing the Process. This Process will either run synchronously or asynchronously.

In the **production code**, it’s desirable that this happens asynchronously because we don’t want the user to wait for his/her REST response until a Process is done with the execution, it’ll return almost immediately and give a “running” or “scheduled” if it’s not done yet, but it’ll continue to run in a different thread and the Process db object will be updated when this background thread is done running.
This way the user won’t ever have to wait ages for the REST reply even when executing large scripts. We don’t want the user to wait or any front-ends to be blocked whilst waiting for a response on the call when it could just be ran in the background and update the db when it’s finished executing.

In the **test code** however, we want to actually test things after a REST call is made. This means that we can’t have processes making alterations to the DB or be stuck in execution whilst the assertions are being tested in the tests. The result that’s coming from the REST api has to be consistent and represent a state in which the Process is fully finished executing, whether it’s completed or failed. This is necessary because otherwise our tests aren’t going to be the same every time they’re ran and they’ll cause failures and inconsistencies.

Implementing this SyncThreadExecutor ensures that any Processes started by the DSpaceRunnableThreadExecutor will have to fully run and be fully done before the REST response is sent. This means that we can check for completed or failed instead of sometimes running into “running” status when the Process would still be executing. We’re simply forcing the Process to fully finish its execution before we construct the REST response so that it’s always consistent and non-variable.

If you remember from a while back, [I mentioned that the tests were failing inconsistently](https://github.com/DSpace/DSpace/pull/2934#issuecomment-682358229), this was mostly because of the executor that was used.
Because of the asynchronous run that was being done at that point in time, the tests did not correctly await the responses/checked the assertions as was described above.
We did removed an unneeded dspaceRunnableThreadExecutor in the api scripts.xml. This is a bean that’s only needed in the server module, so it should be in the REST package only if I’m not horribly mistaken